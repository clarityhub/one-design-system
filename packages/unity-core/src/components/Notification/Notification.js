/** @jsx jsx */
import { jsx, css } from '@emotion/core';

import Typography from '../Typography';
import borders from '../../theme/borders';
import colors from '../../theme/colors';

const base = css`
  border-radius: ${borders.borderRadius};
  padding: 2rem;
  position: relative;
`;

const types = {
	danger: css`
    background-color: ${colors.danger};
    color: ${colors.white};
  `,
	primary: css`
    background-color: ${colors.primary};
    color: ${colors.white};
  `,
	success: css`
    background-color: ${colors.success};
    color: ${colors.white};
  `,
};

const variants = {
	block: css`
    padding: 2rem;
  `,
	thin: css`
    padding: 0.6rem 2rem;
  `,
};

const notificationStyles = ({
	type = 'danger',
	variant = 'block',
}) => css`
  ${base}
  ${types[type]}
  ${variants[variant]}
`;

const Notification = ({ children,  ...props }) => (
	<div
		css={notificationStyles(props)}
		role="alert"
		{...props}
	>
		<Typography
			color="white"
			type="text2"
			noMargin
			noPadding
		>
			{children}
		</Typography>
	</div>
);

export default Notification;
