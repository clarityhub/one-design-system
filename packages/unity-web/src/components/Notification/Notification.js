/** @jsx jsx */
import { jsx, css } from '@emotion/core';
import { node, oneOf, oneOfType, func, string } from 'prop-types';
import borders from '@clarityhub/unity-core/lib/borders';
import colors from '@clarityhub/unity-core/lib/colors';

import Typography from '../Typography';

const base = css`
  border-radius: ${borders.borderRadius.rem}rem;
  padding: 2rem;
  position: relative;
`;

const types = {
	danger: css`
    background-color: ${colors.danger.default};
    color: ${colors.white.default};
  `,
	primary: css`
    background-color: ${colors.primary.default};
    color: ${colors.white.default};
  `,
	success: css`
    background-color: ${colors.success.default};
    color: ${colors.white.default};
  `,
};

const variants = {
	block: css`
    padding: 2rem;
  `,
	thin: css`
    padding: 0.6rem 2rem;
  `,
};

const notificationStyles = ({
	type = 'danger',
	variant = 'block',
}) => css`
  ${base}
  ${types[type]}
  ${variants[variant]}
`;

const Notification = ({
	children,
	component: Component = 'div',
	type = 'danger',
	variant = 'block',
	...props
}) => (
	<Component
		css={notificationStyles({ type, variant })}
		role="alert"
		{...props}
	>
		<Typography
			color="white"
			type="text2"
			noMargin
			noPadding
			component="div"
		>
			{children}
		</Typography>
	</Component>
);

Notification.propTypes = {
	children: node,
	component: oneOfType([func, string]),
	type: oneOf([
		'danger',
		'primary',
		'success',
	]),
	variant: oneOf([
		'block',
		'thin',
	]),
};

export default Notification;
