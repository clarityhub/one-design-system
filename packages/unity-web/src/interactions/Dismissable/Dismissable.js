import React, { Component } from 'react';
import { node, func } from 'prop-types';
import styled from '@emotion/styled';
import noop from '@clarityhub/unity-core/lib/utilities/noop';

const ANIMATION_LENGTH = 300;

const DismissButton = styled.button`
    position: absolute;
    cursor: pointer;
    font-size: 1.2rem;
    top: 0;
    right: 0;
    padding: .75rem 1.25rem;
    color: inherit;

    background-color: transparent;
    border: 0;
    appearance: none;

    /* span {
      background-color: rgba(0, 0, 0, 0.4);
      border-radius: 50%;
      display: inline-block;
      height: 1.4rem;
      width: 1.4rem;
      text-align: center;
    } */
`;

export default class Dismissable extends Component {
  static propTypes = {
  	children: node.isRequired,
  	onDismiss: func,
  }

  static defaultProps = {
  	onDismiss: noop,
  }

  state = {
  	status: 'show',
  }

  componentWillUnmount() {
  	if (this.timer) {
  		clearTimeout(this.timer);
  	}
  }

  onDismiss = (e) => {
  	this.setState({
  		status: 'hiding',
  	}, () => {
  		this.timer = setTimeout(() => {
  			const { onDismiss } = this.props;
  			this.setState({
  				status: 'hidden',
  			}, onDismiss);
  		}, ANIMATION_LENGTH);
  	});
  }

  reset = () => {
  	this.setState({
  		status: 'show',
  	});
  }

  render() {
  	const { children } = this.props;
  	const { status } = this.state;

  	const Dismiss = () => (
  		<DismissButton
  			onClick={this.onDismiss}
  			type="button"
  			aria-label="Dismiss"
  		>
  			<span aria-hidden="true">Ã—</span>
  		</DismissButton>
  	);

  	if (status === 'hidden') {
  		return null;
  	}

  	return children({
  		onDismiss: this.onDismiss,
  		Dismiss,
  	});
  }
}