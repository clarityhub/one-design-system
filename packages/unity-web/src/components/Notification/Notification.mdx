---
name: Notification
menu: Components
---

import { Playground, PropsTable } from 'docz';
import Notification from './Notification';
import Dismissable from '../../interactions/Dismissable';

# Notification

## \<Notification /\> component

```js
import Notification from '@clarityhub/unity-web/libcomponents/Notification';
```

## Prop Types

<PropsTable of={Notification} />

## Examples

<Playground>
  <Notification type="danger" variant="block">
      There were problems with processing your
      request. We encountered the following issues:
      <ul>
          <li>Our servers are offline</li>
      </ul>
  </Notification>
  <br />
  <Notification type="primary" variant="thin">
      Welcome to Clarity Hub, we've updated our Terms and Conditions
  </Notification>
  <br />
  <Notification type="success" variant="block">
      Congratulations, you've enabled your first integration!
  </Notification>
</Playground>

### Dismissable

<Playground>
  <Dismissable>
    {({ Dismiss }) => (
      <Notification
        type="danger"
        variant="block"
      >
        <Dismiss />
        There were problems with processing your
        request. We encountered the following issues:
        <ul>
            <li>Our servers are offline</li>
        </ul>
      </Notification>
    )}
  </Dismissable>
</Playground>

### Autodismiss
