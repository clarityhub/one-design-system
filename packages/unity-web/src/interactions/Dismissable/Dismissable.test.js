import React from 'react';
import renderer, { act } from 'react-test-renderer';

import Notification from '../../components/Notification';
import Dismissable from './Dismissable';

describe('<Dismissable />', () => {
	it('renders text', () => {
		const node = (
			<Dismissable>
				{({ Dismiss }) => (
					<Notification
						type="danger"
						variant="block"
					>
						<Dismiss />
        There were problems with processing your
        request. We encountered the following issues:
						<ul>
							<li>Our servers are offline</li>
						</ul>
					</Notification>
				)}
			</Dismissable>
		);

		const tree = renderer.create(node).toJSON();

		expect(tree).toMatchSnapshot();
	});
    
	it('dismisses the notification', () => {
		const node = (
			<Dismissable dismissImmediately>
				{({ Dismiss }) => (
					<Notification
						type="danger"
						variant="block"
					>
						<Dismiss />
        There were problems with processing your
        request. We encountered the following issues:
						<ul>
							<li>Our servers are offline</li>
						</ul>
					</Notification>
				)}
			</Dismissable>
		);

		const component = renderer.create(node);
        
		act(() => component.root.findByType('button').props.onClick(new Event('click')));

		const tree = component.toJSON();

		expect(tree).toMatchSnapshot();
	});
});