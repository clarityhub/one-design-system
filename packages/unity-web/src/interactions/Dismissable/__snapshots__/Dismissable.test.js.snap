// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dismissable /> dismisses the notification 1`] = `null`;

exports[`<Dismissable /> renders text 1`] = `
<div
  className="css-5swc56-base-danger-block"
  role="alert"
>
  <div
    className="css-d4n4rw"
  >
    <button
      aria-label="Dismiss"
      className="css-mpjvgz-DismissButton e1jp307v0"
      onClick={[Function]}
      type="button"
    >
      <span
        aria-hidden="true"
      >
        Ã—
      </span>
    </button>
    There were problems with processing your request. We encountered the following issues:
    <ul>
      <li>
        Our servers are offline
      </li>
    </ul>
  </div>
</div>
`;
